lib/http/HttpClient_Android.cpp

```diff
@@ -4,505 +4,569 @@
-		if (u == request)
-		{
-			u = m_requests.back();
-			m_requests.pop_back();
-			return;
-		}
-	}
-}
-
-HttpClient_Android::HttpRequest* HttpClient_Android::GetAndRemoveRequest(std::string id)
-{
-	std::lock_guard<std::mutex> lock(m_requestsMutex);
-	for (auto&& u : m_requests)
-	{
-		if (u->m_id == id)
-		{
-			auto r = u;
-			u = m_requests.back();
-			m_requests.pop_back();
-			return r;
-		}
-	}
-	return nullptr;
-}
-
-// since std::to_string() is unavailable (for now), old-school
-// string from number.
-
-std::string HttpClient_Android::NextIdString()
-{
-	auto id_binary = (m_id += 1u);
-	constexpr size_t digits = 11;
-	constexpr uint64_t shift = 6;
-	constexpr uint64_t mask = (static_cast<uint64_t>(1) << shift) - 1u;
-	char buffer[digits + 1];
-	size_t i;
-	for (i = 0; id_binary && i < digits; ++i)
-	{
-		auto r = id_binary & mask;
-		buffer[i] = ' ' + r;
-		id_binary >>= shift;
-	}
-	buffer[i] = 0;
-	return std::string(buffer);
-}
+    constexpr static auto Tag = "HttpClient_Android";
+
+    HttpClient_Android::HttpRequest::~HttpRequest() noexcept
+    {
+        EraseFromParent();
+        if (m_java_request)
+        {
+            JNIEnv* env = nullptr;
+            if (HttpClient_Android::s_java_vm->AttachCurrentThread(&env, nullptr) != JNI_OK)
+            {
+                return;
+            }
+            env->DeleteGlobalRef(m_java_request);
+        }
+    }
+
+    void HttpClient_Android::HttpRequest::SetMethod(std::string const& method)
+    {
+        m_method = method;
+    }
+
+    void HttpClient_Android::HttpRequest::SetUrl(std::string const& url)
+    {
+        m_url = url;
+    }
+
+    HttpHeaders& HttpClient_Android::HttpRequest::GetHeaders()
+    {
+        return m_headers;
+    }
+
+    void HttpClient_Android::HttpRequest::SetBody(std::vector<uint8_t>& body)
+    {
+        m_body = body;  // copy assignment for Great Copy Fun
+    }
+
+    std::vector<uint8_t>& HttpClient_Android::HttpRequest::GetBody()
+    {
+        return m_body;
+    }
+
+    void HttpClient_Android::HttpRequest::SetLatency(EventLatency latency)
+    {
+        // tbh, probably not ever going to do anything about this
+    }
+
+    size_t HttpClient_Android::HttpRequest::GetSizeEstimate() const
+    {
+        return m_body.size();  // nope, but what the heck
+    }
+
+    HttpClient_Android::HttpClient_Android()
+    {
+        m_id = 0;
+    }
+
+    HttpClient_Android::~HttpClient_Android()
+    {
+        JNIEnv* env;
+        s_java_vm->AttachCurrentThread(&env, nullptr);
+        env->DeleteGlobalRef(m_client);
+        m_client = nullptr;  // well, why not?
+    }
+
+    IHttpRequest* HttpClient_Android::CreateRequest()
+    {
+        HttpRequest* local_request(new HttpRequest(*this));
+        std::lock_guard<std::mutex> lock(m_requestsMutex);
+        m_requests.push_back(std::move(local_request));
+        return local_request;
+    }
+
+    // we do NOT own the callback object
+    // we will segfault if callback is nullptr
```

Comment:
My theory is that callback == nullptr would represent a logic error in the calling code. I would rather let hardware deal with the joys of this exceptional condition, even if it means that we get a naked SIGSEGV. Debugging SIGSEGV-because-address-zero is not rocket science. Messing with branch prediction to avoid SIGSEGV is a performance hit.